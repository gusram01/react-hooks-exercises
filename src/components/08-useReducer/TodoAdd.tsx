import React from 'react'
import { useForm } from '../../hooks/useForm';
import { TodoState } from './TodoReducer';

interface Props {
  handleAddTodo: (newTodo: TodoState) => void;
}

const TodoAdd: React.FC<Props> = ({ handleAddTodo }) => {
  const { values, handleInputChange, reset } = useForm({
    description: ''
  });

  const { description } = values;

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    if (description!.trim().length <= 2) {
      return;
    }

    const newTodo: TodoState = {
      id: new Date().getTime(),
      desc: description!,
      done: false
    };
    handleAddTodo(newTodo);
    reset();
  }

  return (
    <>
      <h4>Add Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          className="form-control"
          placeholder="Enter todo"
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Add
            </button>
      </form>
    </>
  )
}

export default TodoAdd;
