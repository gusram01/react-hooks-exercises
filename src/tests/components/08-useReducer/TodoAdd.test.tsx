import React from 'react';
import { shallow } from 'enzyme';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe('TodoAdd tests', () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(
    <TodoAdd
      handleAddTodo={handleAddTodo}
    />
  );

  test('TodoAdd should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('shouldnt called handleAddTodo ', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    // type Error, don't affect test
    //@ts-expect-error
    formSubmit({ preventDefault() { } });
    expect(handleAddTodo).not.toHaveBeenCalled();
  });

  test('handleAddTodo should have been called', () => {
    const value = 'Learn React';
    const input = wrapper.find('input');
    input.simulate('change', { target: { value, name: 'description' } });
    const formSubmit = wrapper.find('form').prop('onSubmit');
    // type Error, don't affect test
    //@ts-expect-error
    formSubmit({ preventDefault() { } });
    expect(handleAddTodo).toHaveBeenCalled();
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      done: false,
      desc: value
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  })

});