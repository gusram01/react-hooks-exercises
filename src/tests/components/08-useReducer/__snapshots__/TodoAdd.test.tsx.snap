// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd tests TodoAdd should render correctly 1`] = `
<Fragment>
  <h4>
    Add Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Enter todo"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
    >
      Add
    </button>
  </form>
</Fragment>
`;
